File: frontend/src/api/commerce.ts
import { apiClient } from './client'
import type { AuthUser } from '../stores/authStore'
import type { PayChannel } from '../utils/payEnv'
import { PENDING_ORDER_KEY } from '../utils/payEnv'

export interface GameMatch {
  id: number
  group: string | null
  date: string | null
  time: string | null
  team1: string | null
  team2: string | null
  stadium: string | null
  can_predict?: boolean
  can_cheer?: boolean
  is_main_team?: boolean
  is_sub_team?: boolean
  has_star_player?: boolean
  user_predicted?: boolean
  user_pick?: string | null
  user_prediction_status?: string | null
  user_stake_coins?: number | null
  user_is_free?: boolean
  pick_stats?: PickStats | null
}

export interface GamePrediction {
  id: number
  match_id: number
  pick: string
  pick_label?: string | null
  stake_coins: number
  is_free: boolean
  status: string
  status_label?: string | null
  points_awarded: number
  redeem_points_awarded?: number
  coins_returned: number
  team1?: string | null
  team2?: string | null
  match_date?: string | null
  match_time?: string | null
  final_score?: string | null
  settled_at?: string | null
  created_at?: string | null
}

export interface Product {
  id: number
  sku: string
  name: string
  description: string | null
  price_fen: number
  coins_grant: number
  grant_season_pass_days: number
  product_type: string
  pay_currency?: string
  redeem_price?: number | null
  grant_payload?: Record<string, unknown> | null
  per_user_limit?: number
  featured?: boolean
}

export interface RedeemProduct extends Product {
  stock_total?: number
  stock_sold?: number
  stock_remaining?: number | null
  is_unlimited_stock?: boolean
  is_out_of_stock?: boolean
  user_purchased_count?: number
  can_purchase?: boolean
  purchase_blocked_reason?:
    | 'login_required'
    | 'limit_reached'
    | 'insufficient_points'
    | 'out_of_stock'
    | null
}

export interface RedeemPurchaseResult {
  order: RedeemOrder
  redeem_points_after: number
  stock_remaining: number | null
}

export interface RedeemOrder {
  id: number
  product_id: number
  redeem_price: number
  status: string
  created_at?: string | null
  product_name?: string | null
}

export interface LeaderboardEntry {
  user_id: number
  nickname: string
  season_points?: number
  redeem_points?: number
  points?: number
  rank?: number
  win_streak: number
  win_rate?: number
  wins?: number
  settled?: number
  predict_points?: number
  battalion_points?: number
  tier_label?: string
  is_me?: boolean
}

export interface LeaderboardBoard {
  board: string
  period: string
  period_label: string
  metric: string
  description: string
  min_samples?: number
  rows: LeaderboardEntry[]
}

export interface MyLeaderboardSummary {
  user_id: number
  nickname: string
  season_points: number
  season_rank: number | null
  redeem_points: number
  redeem_rank: number | null
  daily_points: number
  daily_rank: number | null
  redeem_daily_points: number
  redeem_daily_rank: number | null
  weekly_points: number
  weekly_rank: number | null
  redeem_weekly_points: number
  redeem_weekly_rank: number | null
  win_streak: number
  season_gap_to_prev?: number | null
  redeem_gap_to_prev?: number | null
  battalion_points: number
  battalion_rank: number | null
  battalion_team: string | null
  arena_tier: string
  tier_label: string
  predict: { won: number; lost: number; void: number; settled: number; win_rate: number }
  predict_accuracy_rank: number | null
}

export async function getPredictableMatches() {
  const { data } = await apiClient.get<GameMatch[]>('/api/game/matches')
  return data
}

export async function submitPrediction(payload: {
  match_id: number
  pick: string
  stake_coins?: number
  use_free?: boolean
}) {
  const { data } = await apiClient.post<GamePrediction>('/api/game/predict', payload)
  return data
}

export async function raisePredictionStake(matchId: number, additionalStakeCoins: number) {
  const { data } = await apiClient.post<GamePrediction>('/api/game/predict/raise-stake', {
    match_id: matchId,
    additional_stake_coins: additionalStakeCoins,
  })
  return data
}

export async function getMyPredictions() {
  const { data } = await apiClient.get<GamePrediction[]>('/api/game/my-predictions')
  return data
}

export async function signin() {
  const { data } = await apiClient.post<{
    fan_coins: number
    added: number
    match_day_bonus?: boolean
    signin_streak?: number
    streak_bonus?: number
  }>('/api/game/signin')
  return data
}

export interface DailyStatus {
  signed_today: boolean
  last_signin_date?: string | null
  signin_streak: number
  signin_streak_bonus_next?: number | null
  free_predict: { used: number; limit: number; remaining: number }
  quiz: { answered: boolean }
  pending_predictions: number
  next_pending_match?: { match_id: number; label: string; hours_until?: number | null } | null
  streak_risk?: {
    match_id?: number | null
    label?: string | null
    hours_until?: number | null
    win_streak: number
    streak_bonus_next: number
    message: string
  } | null
  win_streak: number
  loss_streak?: number
  redeem_progress?: {
    next_sku: string
    next_name: string
    need: number
    have: number
    gap: number
    pct: number
  } | null
  match_day: boolean
  match_day_message?: string | null
  free_cheer_tickets?: number
  loss_streak_protect_after?: number
  loss_streak_multiplier?: number
  checklist?: Array<{
    key: string
    label: string
    done: boolean
    reward: string
    optional?: boolean
  }>
  next_action?: {
    key: string
    label: string
    path?: string | null
    hint?: string | null
  }
  ritual_progress?: { done: number; total: number; pct: number }
}

export interface PickStats {
  match_id: number
  total: number
  distribution: Record<string, { count: number; pct: number }>
}

export interface PredictPreview {
  pick: string
  use_free: boolean
  stake_coins: number
  on_win: {
    season_points: number
    redeem_points: number
    coins_returned: number
    free_win_coins: number
    streak_bonus: number
  }
}

export interface WinFeedItem {
  nickname: string
  team1: string
  team2: string
  points_awarded: number
  settled_at?: string | null
}

export async function getDailyStatus() {
  const { data } = await apiClient.get<DailyStatus>('/api/game/daily-status')
  return data
}

export async function getMatchPickStats(matchId: number) {
  const { data } = await apiClient.get<PickStats>(`/api/game/matches/${matchId}/pick-stats`)
  return data
}

export async function getPredictPreview(params: {
  pick: string
  stake_coins?: number
  use_free?: boolean
}) {
  const { data } = await apiClient.get<PredictPreview>('/api/game/predict/preview', { params })
  return data
}

export async function getWinFeed(limit = 15) {
  const { data } = await apiClient.get<WinFeedItem[]>('/api/game/win-feed', { params: { limit } })
  return data
}

export async function getLeaderboard(period: string) {
  const { data } = await apiClient.get<LeaderboardEntry[]>('/api/game/leaderboard', { params: { period } })
  return data
}

export async function getLeaderboardBoard(params: {
  board?: string
  period?: string
  team_id?: number
  limit?: number
}) {
  const { data } = await apiClient.get<LeaderboardBoard>('/api/game/leaderboard/board', { params })
  return data
}

export async function getMyLeaderboardSummary() {
  const { data } = await apiClient.get<MyLeaderboardSummary>('/api/game/leaderboard/me')
  return data
}

export async function getLeaderboardRules() {
  const { data } = await apiClient.get('/api/game/leaderboard/rules')
  return data
}

export async function getFanRank() {
  const { data } = await apiClient.get<{ team: string; fans: number }[]>('/api/game/fan-rank')
  return data
}

export async function getProducts() {
  const { data } = await apiClient.get<Product[]>('/api/shop/products')
  return data
}

export async function getRedeemProducts() {
  const { data } = await apiClient.get<RedeemProduct[]>('/api/shop/redeem/products')
  return data
}

export async function redeemPurchase(productId: number, idempotencyKey?: string) {
  const key =
    idempotencyKey ??
    (typeof crypto !== 'undefined' && crypto.randomUUID
      ? crypto.randomUUID()
      : `redeem-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`)
  const { data } = await apiClient.post<RedeemPurchaseResult>('/api/shop/redeem/purchase', {
    product_id: productId,
    idempotency_key: key,
  })
  return data
}

export async function getRedeemOrders() {
  const { data } = await apiClient.get<RedeemOrder[]>('/api/shop/redeem/orders')
  return data
}

export interface RedeemShopRules {
  economy: {
    season_points_label: string
    redeem_points_label: string
    season_points_desc: string
    redeem_points_desc: string
    predict_win_redeem_ratio: number
    no_loss_reward: boolean
  }
  products: Array<{
    sku: string
    name: string
    redeem_price: number | null
    per_user_limit: number
    grant_payload: Record<string, unknown> | null
  }>
  grant_payload_schema?: Record<
    string,
    { type: string; description: string; example: string; values: string[] | null }
  >
  catalog_source?: string
  sync_cli?: string
  sync_api?: string
  catalog_preview?: Array<Record<string, unknown>>
}

export async function getRedeemRules() {
  const { data } = await apiClient.get<RedeemShopRules>('/api/shop/redeem/rules')
  return data
}

export interface PointLedgerEntry {
  id: number
  delta: number
  balance_after: number
  reason: string
  point_bucket: string
  ref_type: string | null
  ref_id: number | null
  created_at: string | null
}

export async function getPointLedger(bucket: 'season' | 'redeem', limit = 30) {
  const { data } = await apiClient.get<{ status: string; bucket: string; data: PointLedgerEntry[] }>(
    '/api/wallet/point-ledger',
    { params: { bucket, limit } },
  )
  return data.data
}

export interface OrderDetail {
  id: number
  out_trade_no: string
  amount_fen: number
  status: string
  paid_at: string | null
  product_name: string
  product_type: string
  coins_grant: number
  grant_season_pass_days: number
  grant_summary?: string[]
  alipay_trade_no: string | null
}

export async function createOrder(
  productId: number,
  ageConfirmed = true,
  payChannel: 'auto' | PayChannel = 'auto',
) {
  const { data } = await apiClient.post<{
    order: { id: number; out_trade_no: string; status: string }
    pay_url: string
    pay_channel: PayChannel
  }>('/api/pay/alipay/create', {
    product_id: productId,
    age_confirmed: ageConfirmed,
    pay_channel: payChannel,
  })
  return data
}

export async function mockPay(outTradeNo: string): Promise<OrderDetail> {
  const { data } = await apiClient.post<OrderDetail>(
    `/api/pay/alipay/mock-pay?out_trade_no=${encodeURIComponent(outTradeNo)}`,
  )
  return data
}

export async function getOrderByTradeNo(outTradeNo: string): Promise<OrderDetail> {
  const { data } = await apiClient.get<OrderDetail>(
    `/api/pay/orders/by-no/${encodeURIComponent(outTradeNo)}`,
  )
  return data
}

/** Ask backend to query Alipay when async notify may have failed. */
export async function syncAlipayOrder(outTradeNo: string): Promise<OrderDetail> {
  const { data } = await apiClient.post<OrderDetail>(
    `/api/pay/alipay/sync?out_trade_no=${encodeURIComponent(outTradeNo)}`,
  )
  return data
}

/** Read cached pending order and return detail if already paid. */
export async function fetchPaidPendingOrder(): Promise<OrderDetail | null> {
  if (typeof sessionStorage === 'undefined') return null
  const pendingNo = sessionStorage.getItem(PENDING_ORDER_KEY)
  if (!pendingNo) return null
  return resolvePaidOrder(pendingNo)
}

/** Try sync + DB read until we know the order is paid (post-Alipay return). */
export async function resolvePaidOrder(outTradeNo: string): Promise<OrderDetail | null> {
  try {
    const synced = await syncAlipayOrder(outTradeNo)
    if (synced.status === 'paid') return synced
  } catch {
    /* notify may still be in flight */
  }
  try {
    const detail = await getOrderByTradeNo(outTradeNo)
    if (detail.status === 'paid') return detail
  } catch {
    /* ignore */
  }
  return null
}

export type PollOrderResult =
  | { ok: true; order: OrderDetail }
  | { ok: false; reason: 'timeout' | 'failed'; order?: OrderDetail }

export async function pollOrderUntilPaid(
  outTradeNo: string,
  opts: { intervalMs?: number; timeoutMs?: number } = {},
): Promise<PollOrderResult> {
  const intervalMs = opts.intervalMs ?? 1500
  const timeoutMs = opts.timeoutMs ?? 45_000
  const deadline = Date.now() + timeoutMs
  let last: OrderDetail | undefined
  let polls = 0

  const paid = await resolvePaidOrder(outTradeNo)
  if (paid) return { ok: true, order: paid }

  while (Date.now() < deadline) {
    if (polls > 0 && polls % 4 === 0) {
      const synced = await resolvePaidOrder(outTradeNo)
      if (synced) return { ok: true, order: synced }
    }
    polls += 1
    last = await getOrderByTradeNo(outTradeNo)
    if (last.status === 'paid') return { ok: true, order: last }
    if (last.status === 'failed' || last.status === 'cancelled') {
      return { ok: false, reason: 'failed', order: last }
    }
    await new Promise((r) => setTimeout(r, intervalMs))
  }

  const final = await resolvePaidOrder(outTradeNo)
  if (final) return { ok: true, order: final }

  if (last?.status === 'pending') return { ok: false, reason: 'timeout', order: last }
  return { ok: false, reason: 'timeout', order: last }
}

export async function getOrder(orderId: number) {
  const { data } = await apiClient.get(`/api/pay/orders/${orderId}`)
  return data
}

export async function updateProfile(payload: { nickname?: string; favorite_team_id?: number | null }) {
  const { data } = await apiClient.patch<AuthUser>('/api/auth/me', payload)
  return data
}

export interface CoinLedgerEntry {
  id: number
  delta: number
  balance_after: number
  reason: string
  ref_type: string | null
  ref_id: number | null
  created_at: string | null
}

export async function getWalletLedger(limit = 50) {
  const { data } = await apiClient.get<{ status: string; data: CoinLedgerEntry[] }>(
    '/api/wallet/ledger',
    { params: { limit } },
  )
  return data.data
}

File: backend/app/api/schemas/commerce.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class SendCodeRequest(BaseModel):
    email: EmailStr
    age_confirmed: bool = Field(description="用户确认已满18周岁并同意协议")


class VerifyCodeRequest(BaseModel):
    email: EmailStr
    code: str = Field(min_length=4, max_length=8)
    age_confirmed: bool = Field(description="用户确认已满18周岁并同意协议")
    invite_code: str | None = Field(default=None, max_length=12)


class RefreshTokenRequest(BaseModel):
    refresh_token: str


class UserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    email: str
    nickname: str
    avatar_url: str | None = None
    avatar_frame: str | None = None
    theme_key: str | None = None
    fan_coins: int
    season_points: int
    redeem_points: int = 0
    extra_free_predict_daily: int = 0
    win_streak: int
    level: int
    favorite_team_id: int | None = None
    secondary_team_id: int | None = None
    profile_completed: bool = False
    fan_cheers_total: int = 0
    fan_level: int = 1
    battalion_points_season: int = 0
    arena_tier: str = "rookie"
    has_season_pass: bool
    has_active_season_pass: bool = False
    season_pass_until: datetime | None = None
    last_signin_date: date | None = None
    signin_streak: int = 0


class AuthTokenResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    user: UserOut
    is_new: bool = False
    referral: dict | None = None


class UpdateProfileRequest(BaseModel):
    nickname: str | None = Field(default=None, max_length=100)
    favorite_team_id: int | None = None


class PredictSubmitRequest(BaseModel):
    match_id: int
    pick: str
    stake_coins: int = 0
    use_free: bool = False


class PredictRaiseStakeRequest(BaseModel):
    match_id: int
    additional_stake_coins: int = Field(ge=1, description="追加质押球迷币数量")


class GamePredictionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    match_id: int
    pick: str
    stake_coins: int
    is_free: bool
    status: str
    points_awarded: int
    redeem_points_awarded: int = 0
    coins_returned: int
    pick_label: str | None = None
    status_label: str | None = None
    team1: str | None = None
    team2: str | None = None
    match_date: str | None = None
    match_time: str | None = None
    final_score: str | None = None
    settled_at: datetime | None = None
    created_at: datetime | None = None


class GameMatchCardOut(BaseModel):
    id: int
    group: str | None = None
    date: str | None = None
    time: str | None = None
    team1: str | None = None
    team2: str | None = None
    stadium: str | None = None
    user_predicted: bool = False
    user_pick: str | None = None
    user_prediction_status: str | None = None
    user_stake_coins: int | None = None
    user_is_free: bool = False


class UserNotificationOut(BaseModel):
    id: int
    category: str
    title: str
    body: str
    payload: dict | None = None
    read_at: datetime | None = None
    created_at: datetime | None = None


class MarkNotificationsReadRequest(BaseModel):
    ids: list[int] | None = None


class LeaderboardRowOut(BaseModel):
    rank: int | None = None
    user_id: int
    nickname: str
    points: int | None = None
    season_points: int | None = None
    redeem_points: int | None = None
    battalion_points: int | None = None
    win_streak: int = 0
    win_rate: float | None = None
    wins: int | None = None
    settled: int | None = None
    predict_points: int | None = None
    arena_tier: str | None = None
    tier_label: str | None = None
    is_me: bool = False


class LeaderboardBoardOut(BaseModel):
    board: str
    period: str
    period_label: str
    metric: str
    description: str
    min_samples: int | None = None
    rows: list[LeaderboardRowOut]


class MyLeaderboardSummaryOut(BaseModel):
    user_id: int
    nickname: str
    season_points: int
    season_rank: int | None = None
    season_gap_to_prev: int | None = None
    redeem_points: int = 0
    redeem_rank: int | None = None
    redeem_gap_to_prev: int | None = None
    daily_points: int
    daily_rank: int | None = None
    redeem_daily_points: int = 0
    redeem_daily_rank: int | None = None
    weekly_points: int
    weekly_rank: int | None = None
    redeem_weekly_points: int = 0
    redeem_weekly_rank: int | None = None
    win_streak: int
    battalion_points: int
    battalion_rank: int | None = None
    battalion_team: str | None = None
    arena_tier: str
    tier_label: str
    predict: dict
    predict_accuracy_rank: int | None = None


class ProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sku: str
    name: str
    description: str | None
    price_fen: int
    coins_grant: int
    grant_season_pass_days: int
    product_type: str
    pay_currency: str = "cash"
    redeem_price: int | None = None
    grant_payload: dict | None = None
    per_user_limit: int = 0
    stock_total: int = 0
    stock_sold: int = 0
    featured: bool = False


class RedeemProductOut(ProductOut):
    user_purchased_count: int = 0
    stock_remaining: int | None = None
    is_unlimited_stock: bool = True
    is_out_of_stock: bool = False
    can_purchase: bool = False
    purchase_blocked_reason: str | None = None


class RedeemPurchaseRequest(BaseModel):
    product_id: int
    idempotency_key: str | None = Field(default=None, max_length=64)


class RedeemOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    redeem_price: int
    status: str
    created_at: datetime | None = None
    product_name: str | None = None


class RedeemPurchaseOut(BaseModel):
    order: RedeemOrderOut
    redeem_points_after: int
    stock_remaining: int | None = None


class RedeemProductAdminCreate(BaseModel):
    sku: str = Field(min_length=1, max_length=50)
    name: str = Field(min_length=1, max_length=100)
    description: str | None = None
    redeem_price: int = Field(gt=0)
    grant_payload: dict
    per_user_limit: int = Field(default=0, ge=0)
    stock_total: int = Field(default=0, ge=0)
    sort_order: int = 0
    featured: bool = False
    active: bool = True


class RedeemProductAdminUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None
    redeem_price: int | None = Field(default=None, gt=0)
    grant_payload: dict | None = None
    per_user_limit: int | None = Field(default=None, ge=0)
    stock_total: int | None = Field(default=None, ge=0)
    sort_order: int | None = None
    featured: bool | None = None
    active: bool | None = None


class RedeemProductAdminOut(ProductOut):
    active: bool
    stock_remaining: int | None = None
    updated_at: datetime | None = None


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    out_trade_no: str
    amount_fen: int
    status: str
    paid_at: datetime | None = None


class OrderDetailOut(OrderOut):
    product_name: str
    product_type: str
    coins_grant: int
    grant_season_pass_days: int
    alipay_trade_no: str | None = None
    grant_summary: list[str] = Field(default_factory=list)


class CreateOrderRequest(BaseModel):
    product_id: int
    age_confirmed: bool = Field(description="用户确认已满18周岁并知悉虚拟商品规则")
    pay_channel: Literal["auto", "page", "wap"] = "auto"


class CreatePayResponse(BaseModel):
    order: OrderOut
    pay_url: str
    pay_channel: Literal["page", "wap"]
